# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title find all mapped region in a `.sam` data.frame
#' @name sam_to_loc_df
#' 
#' @param sam data.frame. see [read_sam()]
#' 
#' @return a tibble of 4 variables, `chrom`, `strand`, `start`, `end`
#' 
#' @examples
#' sam_file <- system.file('extdata', 'Neat1_1.Aligend_trunc.sam', package = 'paristools')
#' sam_file %>% read_sam() %>% sam_to_loc_df()
#' 
#' @export
NULL

test_parse_CIGAR <- function(chrom, flag, pos, CIGAR, SEQ_len) {
    .Call(`_paristools_test_parse_CIGAR`, chrom, flag, pos, CIGAR, SEQ_len)
}

sam_to_loc_df <- function(sam) {
    .Call(`_paristools_sam_to_loc_df`, sam)
}

#' @title calculate genome coverage for one chromosome
#' @name cal_coverage_impl
#' 
#' @details we use 1-based coordinate, [start, end]
#' 
#' @param chrom_loc_df data.frame. Locations on one chromosome, columns are `strand`, `start`, `end`. There must be at least one region of minimum length 1.
#' 
#' @return tibble. Columns are `pos`, `+`, `-`, the latter two are integers representing the genome, index is 1-based genome position, value is number of reads covering that position.
#' 
#' @keywords internal
NULL

get_strand <- function(flag) {
    .Call(`_paristools_get_strand`, flag)
}

test_as_loc_df <- function() {
    .Call(`_paristools_test_as_loc_df`)
}

cal_coverage_impl <- function(chrom_loc_df) {
    .Call(`_paristools_cal_coverage_impl`, chrom_loc_df)
}

cpp_version <- function() {
    .Call(`_paristools_cpp_version`)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('_paristools_RcppExport_registerCCallable', PACKAGE = 'paristools')
})
